Document the experimental KMIP client API introduced in Vault 2.0.0 (#3239)
hashicorp/web-unified-docs: content/vault/v2.x/content/api-docs/secret/kmip.mdx

    --data @payload.json \
    https://127.0.0.1:8200/v1/kmip/scope/myscope/role/myrole/credential/revoke
```

## Client API (alpha)

~> **Important:** The client API is experimental and may have backwards incompatible
changes introduced in subsequent versions of Vault, or be removed altogether. The client

> API is disabled by default and should **NOT** be used in production.

Reviewer: ```suggestion
@include '../../../global/partials/alerts/status/experimental.mdx'
```
We have an established partial to note experimental features

The client API endpoints can be enabled via the `experiment:kmip.client_api.alpha1`
[experiment option](/vault/docs/configuration#experiments) in the Vault
configuration or on the command-line:

```shell-session
$ vault server -experiment kmip.client_api.alpha1
```

The client API lets you send KMIP operations directly to the Vault KMIP secrets
engine through the Vault HTTP API, without needing a native KMIP client
library. Requests are expressed as XML documents following the [KMIP XML
encoding specification](https://docs.oasis-open.org/kmip/profiles/v1.4/os/kmip-profiles-v1.4-os.html#_Toc491431438).

Vault ships a library of built-in XML request templates for the most common
operations.

### List templates

| Method | Path              |
| :----- | :---------------- |
| `LIST` | `/kmip/template` |

List the names of the built-in XML request templates.

#### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request LIST \

>     https://127.0.0.1:8200/v1/kmip/template

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/kmip/template
```

```

#### Sample response

```json
{
  "data": {
    "keys": [
      "addattribute",
      "create",
      "destroy",
      "get",
      "getattribute",
      "getattributelist",
      "getattributes",
      "locate",
      "locate_addattribute",
      "locate_by_name",
      "locate_decrypt",
      "locate_encrypt",
      "locate_get",
      "locate_getattribute",
      "locate_getattributelist",
      "locate_getattributes",
      "locate_modifyattribute",
      "locate_revoke",
      "locate_revoke_destroy",
      "modifyattribute",
      "revoke",
      "revoke_destroy"
    ]
  }
}
```

### Read template

| Method | Path                    |
| :----- | :---------------------- |
| `GET`  | `/kmip/template/:name` |

Read the XML content of a built-in request template and the list of parameter
placeholders it contains.

#### Parameters

- `name` (`string: <required>`) - Name of the template. This is part of the

>   request URL.

Reviewer: ```suggestion
- `name` (`string: <required>`) - Name of the template.
```

#### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request GET \

>     https://127.0.0.1:8200/v1/kmip/template/create

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/kmip/template/create
```

```

#### Sample response

```json
{
  "data": {
    "template": "...",
    "parameters": ["algorithm", "length", "name", "usage_mask"]
  }
}
```

### Execute KMIP request

| Method | Path                                                            |
| :----- | :-------------------------------------------------------------- |
| `POST` | `/kmip/scope/:scope/role/:role/client[/:operations]` |

Execute one or more KMIP operations on behalf of the given scope and role.
The request is expressed as an XML document. You can provide either:

- A **built-in template name** (e.g. `create`, `locate_encrypt`) and its
  parameters, **or**
- A **custom XML template** string with `PARAMETER_<name>` placeholders and the
  matching parameter values, **or**
- The **operations** path segment (e.g. `.../client/locate/encrypt`) combined
  with parameters so that Vault selects the matching built-in template
  automatically.

> #### Parameters

Reviewer: ```suggestion
#### Path parameters
```

> - `scope` (`string: <required>`) - Name of scope. This is part of the request URL.

Reviewer: ```suggestion
- `scope` (`string: <required>`) - Name of scope.
```

> - `role` (`string: <required>`) - Name of role. This is part of the request URL.

Reviewer: ```suggestion
- `role` (`string: <required>`) - Name of role.
```

- `operations` (`string: ""`) - Slash-separated list of KMIP operation names
  (e.g. `locate/encrypt`). The operation of every batch item in the request must

>   be specified.

Reviewer: ```suggestion

#### Request parameters

- `operations` (`string: ""`) - Slash-separated list of KMIP operation names
  (e.g. `locate/encrypt`). You must specify the operation of every batch item
  in the request.
```

- `template` (`string: ""`) - Either the name of a built-in template (e.g.
  `"create"`) or a raw XML template string with `PARAMETER_<name>` placeholders.

>   If omitted, Vault selects a built-in template based on the `operations` URL

Reviewer: ```suggestion
  If you omit `template`, Vault selects a built-in template based on the `operations` URL
```

  segment.
- `params` (`map<string,string>`) - Arbitrary key/value replacement parameters
  for `PARAMETER_<name>` placeholders in the XML template.

The following named parameters are also accepted as top-level fields and are

> equivalent to setting the same key in `params`:

Reviewer: ```suggestion
Vault also accepts some named parameters as top-level fields.
Providing the following parameters has the same result as setting the
same key in `params`:
```

| Parameter               | Description |
| :---------------------- | :---------- |
| `algorithm`             | Cryptographic algorithm |
| `attr_name`             | Name of the attribute |
| `attr_type`             | Type of the attribute (Boolean, DateTime, Enumeration, Integer, Interval, TextString) |
| `attr_value`            | Value of the attribute |
| `block_cipher_mode`     | Block cipher mode (CBC, GCM, etc.) |
| `data`                  | Data byte string (e.g. `"001a2b34"`) |
| `iv_counter_nonce`      | IV, counter, or nonce |
| `length`                | Cryptographic length |
| `name`                  | Name attribute value |
| `padding_method`        | Padding method (e.g. PKCS5) |
| `revocation_reason_code`| Revocation reason (Unspecified, KeyCompromise, Superseded, etc.) |
| `unique_identifier`     | UniqueIdentifier |
| `usage_mask`            | Cryptographic usage mask |

> #### Sample request — create a symmetric key using a built-in template

Reviewer: ```suggestion
#### Sample request

 The following example creates a symmetric key using a built-in template.
```

> ```shell-session

Reviewer: Why do we have a CLI command in the API docs?

Author: I have removed this example.

$ vault write kmip/scope/myscope/role/myrole/client/create \
    name="mykey" \
    algorithm=AES \
    length=256 \
    usage_mask="Encrypt Decrypt"
```

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     https://127.0.0.1:8200/v1/kmip/scope/myscope/role/myrole/client/create

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/kmip/scope/myscope/role/myrole/client/create
```

```

Where `payload.json` contains:

```json
{
  "name": "mykey",
  "algorithm": "AES",
  "length": "256",
  "usage_mask": "Encrypt Decrypt"
}
```

> #### Sample request — locate and encrypt using a built-in template

Reviewer: ```suggestion
#### Sample request

The following example locates and encrypts a KMIP key using a built-in template.
```

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     https://127.0.0.1:8200/v1/kmip/scope/myscope/role/myrole/client/locate/encrypt

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/kmip/scope/myscope/role/myrole/client/locate/encrypt
```

```

Where `payload.json` contains:

```json
{
  "name": "mykey",
  "block_cipher_mode": "GCM",
  "padding_method": "PKCS5",
  "data": "001a2b34",
  "iv_counter_nonce": "aabb1122"
}
```

> #### Sample request — custom XML template

Reviewer: ```suggestion
#### Sample request

The following example uses a custom XML template.
```

> ```shell-session

Reviewer: Same question re: CLI commands in the API reference

Author: I have removed this example.

$ vault write kmip/scope/myscope/role/myrole/client/locate \
    template=@find.mxl \
    params=altname="kek"
```

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     https://127.0.0.1:8200/v1/kmip/scope/myscope/role/myrole/client/get

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/kmip/scope/myscope/role/myrole/client/get
```
